atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java

```diff
@@ -86,7 +86,28 @@ public HikariCPConnectionManager(ConnectionConfig connConfig) {
     public Connection getConnection() throws SQLException {
         long start = System.currentTimeMillis();
         try {
-            return getConnectionInternal();
+            while (true) {
+                HikariDataSource dataSourcePool = getDataSourcePool();
+                Connection conn = dataSourcePool.getConnection();
+
+                try {
+                    testConnection(conn);
+                } catch (SQLException e) {
+                    log.error("Dropping connection which failed validation", e);
+                    dataSourcePool.evictConnection(conn);
+
+                    if (System.currentTimeMillis() > start + connConfig.getCheckoutTimeout()) {
+                        // it's been long enough that had hikari been
+                        // validating internally it would have given up rather
+                        // than retry
+                        throw e;
+                    }
+
+                    continue;
+                }
+
+                return conn;
```

Comment:
nit: would be clearer to return this in the `try` right after testing and then we don't need the continue.

Reply:
@AmlingPalantir I'm thinking something like https://github.com/palantir/atlasdb/commit/c068a49259a1e59382a51826c3c498c64bb680e3 to make some additional cleanups while we're here and use try-with-resources to ensure closure & exception capture of suppressed exceptions for the cases where datasource or connection does not validate

Reply:
May also want to instead do connection validation on return to pool